registry.go

```diff
@@ -0,0 +1,68 @@
+package multihash
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"hash"
+)
+
+// registry is a simple map which maps a multihash indicator number
+// to a standard golang Hash interface.
+//
+// Multihash indicator numbers are reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+// The keys used in this map must match those reservations.
+//
+// Hashers which are available in the golang stdlib will be registered automatically.
+// Others can be added using the Register function.
+var registry = make(map[uint64]func() hash.Hash)
+
+// Register adds a new hash to the set available from GetHasher and Sum.
+//
+// Register has a global effect and should only be used at package init time to avoid data races.
+//
+// The indicator code should be per the numbers reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+//
+// If Register is called with the same indicator code more than once, the last call wins.
+// In practice, this means that if an application has a strong opinion about what implementation to use for a certain hash
+// (e.g., perhaps they want to override the sha256 implementation to use a special hand-rolled assembly variant
+// rather than the stdlib one which is registered by default),
+// then this can be done by making a Register call with that effect at init time in the application's main package.
+// This should have the desired effect because the root of the import tree has its init time effect last.
+func Register(indicator uint64, hasherFactory func() hash.Hash) {
+	registry[indicator] = hasherFactory
+}
+
+// GetHasher returns a new hash.Hash according to the indicator code number provided.
+//
+// The indicator code should be per the numbers reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+//
+// The actual hashers available are determined by what has been registered.
+// The registry automatically contains those hashers which are available in the golang standard libraries
+// (which includes md5, sha1, sha256, sha384, sha512, and the "identity" mulithash, among others).
+// Other hash implementations can be made available by using the Register function.
+// The 'go-mulithash/register/*' packages can also be imported to gain more common hash functions.
+//
+// If an error is returned, it will be ErrSumNotSupported.
+func GetHasher(indicator uint64) (hash.Hash, error) {
+	factory, exists := registry[indicator]
+	if !exists {
+		return nil, ErrSumNotSupported // REVIEW: it's unfortunate that this error doesn't say what code was missing.  Also "NotSupported" is a bit of a misnomer now.
```

Comment:
Yeah, I think we should add an error type here, and drop `ErrSumNotSupported`.

The `Sum` function actually previously returned its own fresh unique `fmt.Errorf` that contained the number intead of `ErrSumNotSupported`, which seems like pretty strong evidence that there's no significant legacy/compat reason for keeping `ErrSumNotSupported`.

@Stebalien I'm going to call on you to be the final arbiter of this choice.

Reply:
Well it also returns `ErrSumNotSupported`, for example if the code is defined but the hash function isn't.

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/sum.go#L51

For example, keccak-384 or x11 have codes, but no corresponding automatically registered function

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/multihash.go#L135

Reply:
Well, "not supported" is correct because we don't support that hash function. But yeah, we should indicate what isn't supported. I believe we did it this way before because go didn't have a way to check any form of `errors.Is`. Now that we have that, we can define some base error and derive a new error from that.

Reply:
Alright, I've done this, wrapping-style, in cbd218cd90910541ab7a67d6a52ba999716ff2d7 .

I tried to keep it as low impact as possible (e.g. keep an `ErrSumNotSupported` value, by that symbol name).  It works, but I wanna jot some notes that I have tepid feelings about how it turns out: when using wrap like this to provide more info, consumers *still* have to change if they really cared about the previous exact error value, because now they have to use `errors.Is` (as demonstrated by the diff to one of our own tests).

If I'd ditched the goal to keep an `ErrSumNotSupported` value, I probably would've written this by introducing a new struct type to contain the code detail... but in terms of downstream effects, it would be sort of potæto/potato: the exact syntax that consumers would need to decide the "type" of error would be different, but it would still be a diff.

I guess there's really no way to do changes to errors in golang that aren't going to break something (and, we don't really get compiler help detecting it, either, because yay, golang), and this is just life.

In this case I'm hoping it just won't be a big problem in practice, for all the reasons already covered up-thread.